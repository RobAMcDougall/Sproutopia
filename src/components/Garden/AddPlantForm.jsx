import React, { useState, useRef, useEffect } from 'react';

const AddPlantForm = ({ visible, onAddPlant, onCancel }) => {
  const [inputValue, setInputValue] = useState('');
  const [selectedDate, setSelectedDate] = useState('');
  const [filteredOptions, setFilteredOptions] = useState([]);
  const [showOptions, setShowOptions] = useState(false);
  const plantOptions = [
    { id: 1, plant_id: 1, name: 'Tomato1', iconUrl: 'https://i.ibb.co/3BbZpgb/tomato.png' },
    { id: 2, plant_id: 1, name: 'Tomato2', iconUrl: 'https://i.ibb.co/3BbZpgb/tomato.png' },
    { id: 3, plant_id: 2, name: 'Carrot', iconUrl: 'https://i.ibb.co/mqrPkQb/carrot.png' },
    { id: 4, plant_id: 3, name: 'Spring Onion', iconUrl: 'https://i.ibb.co/PZb5WHC/springonion.png' },
  ];

  const optionsRef = useRef(null);

  useEffect(() => {
    const handleClickOutside = (event) => {
      if (optionsRef.current && !optionsRef.current.contains(event.target)) {
        setShowOptions(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);

  const handleInputChange = (e) => {
    const value = e.target.value;
    setInputValue(value);
    setFilteredOptions(
      plantOptions.filter(option =>
        option.name.toLowerCase().includes(value.toLowerCase())
      )
    );
    setShowOptions(true);
  };

  const handleDateChange = (e) => {
    setSelectedDate(e.target.value);
  };

  const handleOptionClick = (option) => {
    setInputValue(option.name);
    setShowOptions(false);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const selectedPlant = filteredOptions.find(option => option.name.toLowerCase() === inputValue.toLowerCase());
    if (selectedPlant) {
      onAddPlant({ ...selectedPlant, plantedDate: selectedDate });
      setInputValue('');
      setSelectedDate('');
      setFilteredOptions([]);
    }
  };

  const handleCancel = () => {
    onCancel();
    setInputValue('');
    setSelectedDate('');
    setFilteredOptions([]);
  };

  return (
    visible && (
      <div className="popup-container">
        <div className="popup-content">
          <form onSubmit={handleSubmit}>
            <div className='form-title'>
              Add a plant to your pot!
            </div>
            <div className="input-container">
              <label htmlFor="plant-input" className="input-label">
                Search for a plant:
              </label>
              <div className="search-container">
                <input
                  id="plant-input"
                  type="text"
                  value={inputValue}
                  onChange={handleInputChange}
                  className="search-input"
                />
                {showOptions && filteredOptions.length > 0 && (
                  <div ref={optionsRef} className="options-container">
                    {filteredOptions.map((option, index) => (
                      <div
                        key={index}
                        className="option"
                        onClick={() => handleOptionClick(option)}
                      >
                        {option.name}
                      </div>
                    ))}
                  </div>
                )}
              </div>
            </div>
            <div className="input-container">
              <label htmlFor="date-input" className="input-label">
                Date planted:
              </label>
              <input
                id="date-input"
                type="date"
                value={selectedDate}
                onChange={handleDateChange}
                className="date-input"
              />
            </div>
            <div className="button-container">
              <button className="submit-button">Add</button>
              <button className="cancel-button" onClick={handleCancel}>Cancel</button>
            </div>
          </form>
        </div>
      </div>
    )
  );
};

export default AddPlantForm;
