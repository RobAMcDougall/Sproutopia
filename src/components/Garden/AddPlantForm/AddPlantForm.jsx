import React, { useState, useRef, useEffect } from 'react';

const AddPlantForm = ({ visible, onAddPlant, onCancel }) => {
  const [inputValue, setInputValue] = useState('');
  const [selectedDate, setSelectedDate] = useState('');
  const [filteredOptions, setFilteredOptions] = useState([]);
  const [showOptions, setShowOptions] = useState(false);
  const [plantOptions, setPlantOptions] = useState([]);
  const optionsRef = useRef(null);

  useEffect(() => {
    fetchPlants();
  }, []);

  useEffect(() => {
    const handleClickOutside = (event) => {
      if (optionsRef.current && !optionsRef.current.contains(event.target)) {
        setShowOptions(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);

  const fetchPlants = async () => {
    try {
      const response = await fetch('http://localhost:3000/plants/all');
      if (!response.ok) {
        throw new Error('Failed to fetch plants');
      }
      const plants = await response.json();
      setPlantOptions(plants);
    } catch (error) {
      console.error('Error fetching plants:', error);
    }
  };

  const handleInputChange = (e) => {
    const value = e.target.value;
    setInputValue(value);
    setFilteredOptions(
      plantOptions.filter(option =>
        option.name.toLowerCase().includes(value.toLowerCase())
      )
    );
    setShowOptions(true);
  };

  const handleDateChange = (e) => {
    const date = e.target.value;
    setSelectedDate(date);
  };

  const handleOptionClick = (option) => {
    setInputValue(option.name);
    setShowOptions(false);
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const selectedPlant = filteredOptions.find(option => option.name.toLowerCase() === inputValue.toLowerCase());
    if (selectedPlant) {
      try {
        await onAddPlant(selectedPlant, selectedDate);
        setInputValue('');
        setSelectedDate('');
        setFilteredOptions([]);
      } catch (error) {
        console.error('Error adding plant:', error);
      }
    }
  };

  const handleCancel = () => {
    onCancel();
    setInputValue('');
    setSelectedDate('');
    setFilteredOptions([]);
  };

  return (
    visible && (
      <div className="popup-container">
        <div className="popup-content">
          <form onSubmit={handleSubmit}>
            <div className='form-title'>
              Add a plant to your pot!
            </div>
            <div className="input-container">
              <label htmlFor="plant-input" className="input-label">
                Search for a plant:
              </label>
              <div className="search-container">
                <input
                  id="plant-input"
                  type="text"
                  value={inputValue}
                  onChange={handleInputChange}
                  className="search-input"
                />
                {showOptions && filteredOptions.length > 0 && (
                  <div ref={optionsRef} className="options-container">
                    {filteredOptions.map((option, index) => (
                      <div
                        key={index}
                        className="option"
                        onClick={() => handleOptionClick(option)}
                      >
                        {option.name}
                      </div>
                    ))}
                  </div>
                )}
              </div>
            </div>
            <div className="input-container">
              <label htmlFor="date-input" className="input-label">
                Date planted:
              </label>
              <input
                id="date-input"
                type="date"
                value={selectedDate}
                onChange={handleDateChange}
                className="date-input"
              />
            </div>
            <div className="button-container">
              <button className="submit-button">Add</button>
              <button className="cancel-button" onClick={handleCancel}>Cancel</button>
            </div>
          </form>
        </div>
      </div>
    )
  );
};

export default AddPlantForm;
