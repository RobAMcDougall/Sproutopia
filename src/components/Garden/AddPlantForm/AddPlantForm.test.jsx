import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import AddPlantForm from './AddPlantForm';

describe('AddPlantForm', () => {
  it('renders without crashing', () => {
    render(<AddPlantForm />);
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<AddPlantForm />);
    expect(asFragment()).toMatchSnapshot();
  });
  

});
