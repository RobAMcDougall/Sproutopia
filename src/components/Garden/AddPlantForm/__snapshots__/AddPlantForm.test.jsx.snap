// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddPlantForm > AddPlantForm matches snapshot 1`] = `<DocumentFragment />`;

exports[`AddPlantForm > matches snapshot 1`] = `<DocumentFragment />`;

exports[`AddPlantForm matches snapshot 1`] = `<DocumentFragment />`;
